src/serde/tree_cache.rs

```diff
@@ -0,0 +1,744 @@
+    parents: Vec<(u32, ChildPos)>,
+    /// if this node doesn't have a tree_hash, the serialized length is not
+    /// valid as it cannot be computed.
+    serialized_length: u64,
+    /// set to non-zero, if this node is pushed onto the parse stack. Since a
+    /// node can be on the stack multiple times, it's counted up every time it's
+    /// pushed, and counted down every time it's popped.
+    pub on_stack: u32,
+}
+
+struct PartialPath {
+    // the path we've built so far
+    path: PathBuilder,
+    // if we're traversing the stack, this is the stack position. Note that this
+    // is not an index into the stack array, it's a counter of how far away from
+    // the top of the stack we are. 0 means we're at the top, and we've found
+    // a path.
+    // -1 means we're not traversing the stack.
+    // TODO: these fields are mutually exclusive (stack_pos vs. idx and child).
+    // it might be nice to use an enum, but it's not obvious that it's simpler
+    // or faster
+    stack_pos: i32,
+    // if we're not traversing the stack, this is the next parent, and whether
+    // we're coming from the left or right child
+    idx: u32,
+    child: ChildPos,
+}
+
+enum CacheOp {
+    // node to traverse
+    Traverse(NodePtr),
+    // pair node
+    Cons(NodePtr),
+}
+
+#[derive(Clone)]
+pub struct TreeUndoState {
+    stack: Vec<u32>,
+    lookup: VisitedNodes,
+    sentinel_entry: Option<u32>,
+}
+
+#[derive(Default)]
+pub struct TreeCache {
+    /// caches extra metadata about a tree of nodes. The value is an index into
+    /// the node_entry vector.
+    cache: HashMap<NodePtr, u32>,
+
+    /// The metadata for all nodes in the tree. This is like a shadow tree
+    /// structure to the NodePtr one. The most important difference is that
+    /// identical nodes are merged, using the same NodeEntry, and additional
+    /// metadata is kept, such as the tree hash.
+    node_entry: Vec<NodeEntry>,
+
+    /// maps tree-hashes to the index of the corresponding NodeEntry in the
+    /// node_entry vector. For any given tree hash, we're only supposed to
+    /// have a single NodeEntry. There may be multiple NodePtr referring to
+    /// the same NodeEntry (if they are identical sub trees).
+    hash_to_node: HashMap<Bytes32, u32, RandomState>,
+
+    /// When deserializing, we keep a stack of tokens we've parsed so far, this
+    /// stack is maintaining that same state, since that's what back-references
+    /// are pointing into. Nodes that have the sentinel as direct decendants
+    /// don't have a tree-hash, so it's set to None for those entries
+    stack: Vec<u32>,
+
+    /// This records which NodeEntries have been serialized so far. When we look
+    /// for back-references, we can only pick nodes in this set. nodes with
+    /// small serialized length are not inserted. This lookup is built and
+    /// updated as we serialize, to ensure we only include nodes that *can* be
+    /// referenced.
+    lookup: VisitedNodes,
```

Comment:
I didn't understand at first why this was necessary since it didn't seem like nodes that were visited yet could possibly have entries in `hash_to_node`. But then I realized you can call `update` multiple times (I don't think `ObjectCache` allows that) and `hash_to_node` might have leftovers from a prior call. Am I getting this right?

Reply:
The main difference is that `ObjectCache` builds the tree, and computes tree-hashes and serialized-lengths *as you go*. So, nodes that haven't been serialized yet won't exist in there to be found. With `TreeCache`, the entire tree is traversed in full up-front. When we start to serialize it all nodes can be found in there, including ones that haven't been serialized yet. That's why we need to track this. I'll change the name to something better.

Reply:
I'm struggling to see why this a member variable. Is it because of checkpointing/undo?

Reply:
no, it must persist across calls to `find_path()`, `push()` and `pop2_and_cons()`. It's updated on `push()` (as that's when nodes become available to be found) and used by `find_path()`.